File: 2020/fsharp/AOC/Day9.fs
module AOC.Day9

let checkValid preamble wanted = 
    let rec innerTest left right = 
        if right >= Array.length preamble then false
        elif preamble.[left] > wanted then false
        elif (preamble.[left] + preamble.[right]) = wanted then true
        else innerTest left (right+1)
    let rec outerTest left =
        if left > (Array.length preamble - 1) then false
        elif innerTest left (left+1) then true 
        else outerTest (left+1)
    outerTest 0

let part1 (data: int64 []) lenPreamble = 
    let rec loop idx = 
        if idx > Array.length data then invalidArg "data" "No bad number"
        elif not (checkValid data.[(idx - lenPreamble)..(idx-1)] data.[idx]) then data.[idx]
        else loop (idx + 1)
    loop lenPreamble

let part2 data lenPreamble = 
    let target = part1 data lenPreamble
    let rec widenWindow left right = 
        let total = Array.sum data.[left..right]
        if total = target then 
            Some ((Array.min data.[left..right]) + (Array.max data.[left..right]))
        elif total > target then None 
        else widenWindow (left-1) right

    let rec walkLeft right =
        if right < 1 then invalidArg "data" "No summing section"
        else match widenWindow (right-1) right with 
                | Some v -> v 
                | None -> walkLeft (right-1)
    
    walkLeft ((Array.length data) - 1)

File: 2020/julia/AOC/src/day9.jl
module Day9
import AOC.Parsers

function check_validity(number, preamble)
	for adx in 1:length(preamble)
		if preamble[adx] > number 
			continue 
		end
		for bdx in adx:length(preamble)
			if number == preamble[adx] + preamble[bdx]
				return true 
			end
		end
	end
	false
end

function part1(;data=Parsers.day9(), preamble_len=25)
	for idx in (preamble_len + 1):length(data)
		if !check_validity(data[idx], @view data[(idx-preamble_len):(idx-1)])
			return data[idx]
		end
	end
end

function part2(;data=Parsers.day9(), preamble_len=25)
	target_number = part1(data=data, preamble_len=preamble_len)
	for right in length(data)-1:-1:2
		left = right - 1
		total = sum(@view data[left:right])
		while total < target_number 
			left -= 1
			total = sum(@view data[left:right])
		end
		if total == target_number 
			return sum(extrema(@view data[left:right]))
		end
	end
end

end